apps/wallet-mobile/src/features/Receive/common/ReceiveProvider.tsx

```diff
@@ -0,0 +1,83 @@
+import {produce} from 'immer'
+import _ from 'lodash'
+import React from 'react'
+
+export const useReceive = () => React.useContext(ReceiveContext)
+
+export const ReceiveProvider = ({
+  children,
+  initialState,
+  initialCurrentAddress,
+}: {
+  children: React.ReactNode
+  initialState?: Partial<ReceiveState>
+  initialCurrentAddress?: string
+}) => {
+  const [state, dispatch] = React.useReducer(receiveReducer, {
+    ...defaultState,
+    ...initialState,
+  })
+
+  const actions = React.useRef<ReceiveActions>({
+    currentAddressChanged: (address: string) => dispatch({type: ReceiveActionType.CurrentAddressChanged, address}),
+  }).current
+
+  const context = React.useMemo(
+    () => ({
+      ...state,
+      ...actions,
+      defaultAddress: initialCurrentAddress ?? '',
+    }),
+    [state, actions, initialCurrentAddress],
+  )
+
+  return <ReceiveContext.Provider value={context}>{children}</ReceiveContext.Provider>
+}
+
+const receiveReducer = (state: ReceiveState, action: ReceiveAction) => {
+  return produce(state, (draft) => {
+    switch (action.type) {
+      case ReceiveActionType.CurrentAddressChanged:
+        draft.selectedAddress = action.address
+        break
+
+      default:
+        throw new Error(`invalid action`)
+    }
+  })
+}
+
+export type ReceiveType = 'single' | 'multiple'
```

Comment:
it should be part of the wallet, cuz it will affect how the change are processed, WalletAddressDerivationType

Reply:
@stackchain  in this case we should add it in the types package?

Reply:
it's ok to leave it sitting here now, I'm moving to types only after -> extracting as package.